Refactor framework and library detection code.

kmaglione opened the pull request:
Sorry. This is kind of large.

Thoughts for easier review:

The `JetpackCFX.validate` method is mostly code that was in 'jetpack.js', but with a lot of dead code removed.

The 'test_libraries.py` file is mostly code from the very strangely misnamed 'test_libraryblacklist.py' file, but git's rename tracking is not smart enough to notice.

The code in 'metadata/libfetcher.py' and 'metadata/update.py' can be skipped, if you prefer. It won't be run on servers, and we'll probably be changing it locally without submitting PRs immediately for a lot of runs.

There's a related update to Olympia coming to handle out-of-band updates of the metadata JSON and add some integration with the file viewer.

kmaglione commented:
Oh, and fortunately most of the "files changed" are just deletions of old cruft and metadata files.

magopian commented:
This PR is not mergeable anymore, could you please rebase it?

When git isn't able to detect the renaming/moving of files, you can help it, I believe, by using `git mv`. Maybe this would make this PR more readable?

I'm afraid I'm not familiar enough with the codebase (yet?) to review this PR, maybe @mstriemer will have some time to give it a look?

diox commented:
Bug number ? What is the plan exactly here, if you are removing the old lists of files ?

mstriemer commented:
I didn't get to look at all the code but the cleanup looks a lot nicer than the old code, thanks.

I'm assuming the out-of-band changes PR in olympia will allow for `libraries.json` to be updated outside of git. Will this update the `libraries.json` file on the server or is there some other mechanism to get those changes into the validator?

How will the out-of-band changes interact with the checked-in version?

I'm just wondering what the game plan is here so I know what to expect in all of these changes.

kmaglione commented:
Huh. I rebased before pushing. I guess I'll try again.

I did use `git mv`, but unfortunately it doesn't make a difference.

kmaglione commented:
Related bugs: https://bugzil.la/1069570 https://bugzil.la/1013413

Yes, the out-of-band changes will be to update the libraries.json data. The file itself won't be changed. AMO will pass the data directly to the `validate` method, much like it currently does for the application metadata.

magopian commented:
@kmaglione could you please rebase?

kmaglione commented:
Rebased and fixed some conflicts.

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+VALID_VERSION = re.compile(r"^\d[\d.]+$")
+
+
+def filter_re(expr):
```
docstring?

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+def tag_transform(tag, min_version=None, prefix=None):
+    """
+    Generic function to transform a revision control tag to a version number.
+    Returns `None` for invalid tags.
```
Give an example of what that does (eg, with a prefix)

magopian replied:
And add an explanation of what/why the min_version

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+        identifier for the location of the version in question
+        (usually the URLs of scripts).
+        """
+        return ()
```
`raise NotImplementedError` ?

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+        """
+        raise NotImplementedError
+
+    def get_checksums(self, version, location):
```
not used? shouldn't be used?

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+            yield path, r.content
+
+
+class ZipFetcher(object):
```
Please add `Mixin` to the name: `ZipFetcherMixin`, it helps spotting it ;)

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+    archive, and which optionally implement a `file_filter` method to restrict
+    the checksummed files.
+
+    The entire file is read into memory.
```
I believe this line/comment could be a docstring of the `get_files` method

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+    """
+
+    def get_files(self, version, url):
+        from StringIO import StringIO
```
add a comment why it's not using `cStringIO`

magopian commented on metadata/libfetcher.py:
```diff
@@ -0,0 +1,373 @@
+    def get_files(self, version, url):
+        from StringIO import StringIO
+
+        r = requests.get(url)
```
no one-letter variables please

magopian commented on tests/resources/bootstrap-jpm.js:
```diff
@@ -0,0 +1,385 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This is a test framework used in the tests to make sure it's correctly detected as a framework, right? If so, could it be much shorter, maybe even just containing a line saying that it's a test framework?

In any case, it needs a comment at the beginning of this file to give the context, and explain it's not to be used/updated/...

kmaglione replied:
This is an unmodified file from the add-on SDK. It needs to be included as-is so the stock checksums will match it.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+        self.resources = resources
+
+    def info(self, name):
+        # Um...
```
?

kmaglione replied:
Sorry, the comment was because I don't understand why this method even exists. I'll remove it.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+
+    def read(self, name):
+        if isinstance(self.resources[name], bool):
+            return ""
```
Why an empty string?

magopian replied:
(hint: comment ;)

kmaglione replied:
I moved this from another file. Essentially, whoever implemented it decided that `True` should be used for files which exist but may be empty, rather than just `''`.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+    for elem in path:
+        dict_ = dict_[elem]
+
+    unset = object()
```
Why use this? It doesn't seem idiomatic, and if it's a boolean, why not `dict_.get("banned", False)`?

kmaglione replied:
It's not a boolean. If the property doesn't exist, it needs to be removed entirely after the test, so `unset` is a sentinel. I thought I'd added a comment to that effect.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+
+    yield
+
+    if prev is unset:
```
Oh, ok I understand now (cf precedent comment). I'd use `None` in that case.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+    framework_files = err.metadata.get("framework_files")
+    assert identified_files and framework_files
+
+    for f in files:
```
No one letter variables please

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+            assert (("testcases_frameworks", "banned_framework") in
+                    (e["id"] for e in err.errors))
+        else:
+            # We don't bother reporting framework files for banned
```
That's a comment for the `if` case right?

kmaglione replied:
Yeah. Donno how it wound up in the `else` branch.

magopian commented on tests/test_frameworks.py:
```diff
@@ -0,0 +1,180 @@
+            eq_(set(data["framework_files"]),
+                set(err.metadata["framework_files"]))
+
+            if data.get("unversioned"):
```
You're hardcoding the version below, so I'd also hardcode this branch (for the `besttoolbars` which is the only unversioned one in your fixture).

kmaglione replied:
Hm. The version is generated from the contents of the files. I don't really want to hard code it, since that would rely on the behavior of `fake_version`, or the contents of the stub files, not changing, which I don't want to depend on. I just want to make sure that it comes out to the same as the return value for `fake_version`.

kmaglione replied:
I replaced the hard coded `"42"` with a `VERSION` constant.

magopian commented on tests/test_jetpack.py:
```diff
@@ -5,53 +5,60 @@
-            return ""
-        return self.resources[name]
+    if allow_old_sdk:
+        ignore_ids = [
```
should it be named `ignore_id` (singular)?

kmaglione replied:
If I changed it to a tuple rather than a list, and changed the `in` to `==`. I implemented it this way since it didn't seem to add significant complexity, and it seemed likely we'd want to add other IDs in the future.

magopian commented on tests/test_jetpack.py:
```diff
@@ -5,53 +5,60 @@
 
-    def __contains__(self, name):
-        return name in self.resources
+    print err.print_summary(verbose=True)
```
print?

magopian commented on tests/test_jetpack.py:
```diff
@@ -5,53 +5,60 @@
-    assert not err.errors
-    assert not err.warnings
-    assert not err.notices
+    eq_(err.errors, [])
```
why replace `assert` with `eq_` in this case? Maybe use `assert err.errors == []` if you want a precise comparison?

kmaglione replied:
Because nose doesn't have any special behavior for reporting assert failures, and if this fails, I'd like to see what value it wound up with, without having to add debugging code. I changed this because it was failing, and I needed that information.

magopian replied:
You're right, the validator isn't using pytest (yet? :)

magopian commented on tests/test_jetpack.py:
```diff
@@ -62,111 +69,94 @@ def test_bad_harnessoptions():
     assert err.failed()
-    assert err.warnings
+    assert err.errors
     print err.warnings
```
I would remove the print

magopian commented on tests/test_jetpack.py:
```diff
@@ -62,111 +69,94 @@ def test_bad_harnessoptions():
-                            "bar.JpG": True,
-                            "safe.GIF": True,
-                            "icon.ico": True,
-                            "foo/.DS_Store": True}))
```
You didn't keep the `foo/.DS_Store` file?

kmaglione replied:
No. That file generates a warning, and there didn't seem to be a good reason to keep it.

magopian commented on tests/test_jetpack.py:
```diff
@@ -176,34 +166,32 @@ def test_pass_manifest_elements():
 
-    assert "jetpack_unknown_files" in err.metadata
-    assert not err.metadata["jetpack_unknown_files"]
+    if False:
```
wat

kmaglione replied:
See comment. I can comment out the old lines of code, rather than `if False`ing them, if you prefer.

magopian replied:
Yes please ;) I'd rather have commented out code (even though I dislike it a lot, if it's commented, it's not used, if it's not used, it shouldn't be there) than this construct.

magopian commented on tests/test_jetpack.py:
```diff
@@ -328,20 +300,23 @@ def test_mismatched_db_hash():
-    nose.tools.ok_('resources/bootstrap.js' in unknown_files)
+    # There is currently no analog for this. Restore if we decide to
+    # create one.
+    # assert "jetpack_unknown_files" in err.metadata
```
Do we need this?

kmaglione replied:
I don't think so. I was just leaving it there for reference in case it turns out that someone was using that property.

magopian commented on tests/test_jetpack.py:
```diff
@@ -328,20 +300,23 @@ def test_mismatched_db_hash():
-    print err.print_summary(verbose=True)
-    assert not err.failed()
+                            "harness-options.json": harnessoptions}))
+    assert err.failed()
```
This is the exact opposite as what it was, is that on purpose?

kmaglione replied:
Yes. There was no warning for this in the old code. The new code reports files which are expected to match a known hash but don't.

magopian commented on tests/test_jetpack.py:
```diff
@@ -456,23 +433,12 @@ def test_outdated_sdkversion():
     eq_(err.warnings[0]["id"][2], "outdated_version")
 
-
-def test_future_sdkversion():
```
Don't we want this test anymore?

kmaglione replied:
No. The expected behavior is that we should warn when a future version of the SDK is used. But this test wasn't actually testing that, since 1.8 had not been a future version in a very long time.

magopian commented on tests/test_libraries.py:
```diff
@@ -0,0 +1,128 @@
+
+    test_content.test_packed_packages(err, package)
+
+    print err.print_summary()
```
I don't think we want/need those `print`s

kmaglione replied:
This has been used by the majority of validator tests for a long time. The output doesn't show unless the test fails, and when the test fails, it usually turns out to be useful.

magopian commented on tests/test_libraries.py:
```diff
@@ -0,0 +1,128 @@
+    package = XPIManager(package_data, mode="r", name="blocked.xpi")
+    err = ErrorBundle()
+
+    test_content.test_packed_packages(err, package)
```
I would create a helper function for the 4 lines above that seem to be used throughout this file. Also, I'd use the `with open(...) as ...` idiom.

kmaglione replied:
I agree. I did that with `test_jetpack.py`, I just didn't want to refactor all of these. This file was renamed from `test_libraryblacklist.py`, so most of this code is not new.

magopian commented on tests/test_libraries.py:
```diff
@@ -0,0 +1,128 @@
+        eq_(notice["id"],
+            ("testcases_libraries", "detect_libraries",
+             "detected_library_message"))
+        assert all(m in notice["description"]
```
no one letter variables please

magopian commented on validator/errorbundler.py:
```diff
@@ -384,20 +389,28 @@ def print_summary(self, verbose=False, no_color=False):
+
+            if self.metadata.get("framework_files"):
+                self.handler.write("Framework files:")
+                for filenamename in self.metadata["framework_files"]:
```
typo `filename`

magopian commented on validator/errorbundler.py:
```diff
@@ -384,20 +389,28 @@ def print_summary(self, verbose=False, no_color=False):
+            if self.metadata.get("framework_files"):
+                self.handler.write("Framework files:")
+                for filenamename in self.metadata["framework_files"]:
+                    self.handler.write(" %s" % filenamename)
```
`self.handler.write(' '.join(self.metadata['framework_files']))`

kmaglione replied:
It'd need to be something like `"\n ".join(...)`, but I'm not sure of all of the implications of writing output that way with this code. I was mainly just updating the current code to use the new data structures.

magopian commented on validator/errorbundler.py:
```diff
@@ -384,20 +389,28 @@ def print_summary(self, verbose=False, no_color=False):
+                self.handler.write("Framework files:")
+                for filenamename in self.metadata["framework_files"]:
+                    self.handler.write(" %s" % filenamename)
+                self.handler.write("")
```
what does writing an empty string do?

kmaglione replied:
This function writes newlines after the string. This just writes out a newline.

magopian commented on validator/errorbundler.py:
```diff
@@ -384,20 +389,28 @@ def print_summary(self, verbose=False, no_color=False):
+                for name, data in (self.metadata["identified_files"]
+                                    .iteritems()):
+                    self.handler.write(" %s:" % name)
+                    self.handler.write("\n".join("  %s %s (%s)" % tuple(s)
```
no single letter variable please. What's in `s`? Is the `tuple(s)` call needed?

kmaglione replied:
`s` is a list describing the source. Without the `tuple()`, it's processed as a single argument.

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+    Returns a hash to act in place of a version number, given the contents
+    of a set of known files.
+    """
+    import hashlib
```
import at the top of the module

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+class MetaFramework(type):
+    def __new__(mcls, name, bases, dict_):
+        cls = super(MetaFramework, mcls).__new__(mcls, name, bases, dict_)
+        if cls.NAME:
```
It seems this is only to not list the base `Framework` class, so maybe have a specific case just for it, `if cls.__name__ != 'Framework':`

magopian replied:
Ah, I see you're using the `NAME` property anyway, so I guess it's good this way

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                setattr(self, prop.upper(),
+                        tuple(self.framework_metadata[prop]))
+
+    NAME = None
```
All properties together at the beginning of the class please

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+    @cached_property
+    def ID(self):
+        """Internal ID of the framework."""
+        return self.NAME and self.NAME.lower()
```
This will return `False` if there's no `self.NAME`, and not an empty string or `None`. I understand this is only missing for the base `Framework` class, so maybe you don't need this test?

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+
+    def detect(self):
+        """Returns true if the add-on matches the given framework."""
+        return False
```
`raise NotImplementedError`?

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+        Returns a list of files in the package which belong to the framework
+        in question. Files may or may not match known versions of the file.
+        """
+        return (filter(self.xpi.__contains__, list(self.FRAMEWORK_FILES)) +
```
why the call to `list`, it says from the comment https://github.com/mozilla/amo-validator/pull/271/files#diff-247aeb318317091e4843b00ac0f88b12R60 that it's a tuple?

kmaglione replied:
Because we can't concatenate a tuple to a list, and `filter` will return a tuple if it gets a tuple for input.

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                    return metadata
+            except:
+                pass
+        return None
```
Not needed

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+    @cached_property
+    def version(self):
+        """
+        Returns the version of the framework used, or `None` if not known.
```
I don't think it's returning `None`, but a fake version, right?

kmaglione replied:
Yeah. Behavior changed.

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                             % self.NAME),
+                filename=self.METADATA_FILE)
+
+            return self.valid
```
Shouldn't it set `self.valid = False` first?

magopian replied:
Ah, it's done in `self.error`, my bad

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+            report = self.err.warning if banned == "maybe" else self.error
+            report(
+                err_id=("testcases_frameworks",
+                        "banned_framework"),
```
move on the line above

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+        elif not version_metadata:
+            self.error(
+                err_id=("testcases_frameworks",
+                        "unknown_framework_version"),
```
Does it fit on the previous line?

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+
+                self.error(
+                    err_id=("testcases_frameworks",
+                            "unknown_file"),
```
on the previous line

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                                 "not match any known checksum." % self.NAME),
+                    filename=path)
+
+            elif not any(match[0:2] == tag for match in identity["sources"]):
```
comment

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+        return self.valid
+
+
+# Built on top of Jetpack, so must come first.
```
How do you enforce that? I thought you would have needed to put it before Jetpack in the file?

magopian replied:
Ah, sorry, I understand, the Jetpack class above is an "abstract" one (with `NAME = None`). I would add a docstring.

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                       "chrome/content/crossriderapi.js"))
+
+
+class Jetpack(Framework):
```
Add a docstring explaining that it's an "abstract" (or "base") framework, with `NAME = None`

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+                # Keep track of all of the valid modules that were loaded.
+                loaded_modules.append("".join([module["packageName"], "-",
+                                               module["sectionName"], "/",
+                                               module["moduleName"], ".js"]))
```
This is a strange way to do string interpolation, why not `%s-%s/%s.js`?

kmaglione replied:
Sorry. It's  a JS thing.

kmaglione replied:
Oh, also this is old code from `jetpack.py`

magopian commented on validator/testcases/frameworks.py:
```diff
@@ -0,0 +1,546 @@
+
+    @cached_property
+    def version(self):
+        return self.metadata.get("scriptify-version")
```
you're using `.get()` here, and `[...]` in the previous framework.

kmaglione replied:
Because it's required to be present in Kango, but not in Scriptify.

magopian replied:
understood, thanks

magopian commented on validator/testcases/installrdf.py:
```diff
@@ -30,16 +28,16 @@ def _test_rdf(err, install):
     easier."""
 
-    shouldnt_exist = ("hidden", )
+    shouldnt_exist = "hidden",
```
To be honest, I'd rather `["hidden"]` for perfect readability

kmaglione replied:
I'm going to remove these changes for now. I'm not sure how they wound up in this PR.

magopian commented on validator/testcases/installrdf.py:
```diff
@@ -50,19 +48,18 @@ def _test_rdf(err, install):
                       "multiprocessCompatible",
                       "optionsType",
                       "type",
-                      "updateInfoURL",
```
deprecated?

kmaglione replied:
No, it never actually existed in install.rdf.

magopian commented on validator/testcases/libraries.py:
```diff
@@ -0,0 +1,204 @@
+                    # greater granularity in the future.
+                    banned = merge_banned(data, vdata)
+
+                for path, hash in vdata.get("files", {}).iteritems():
```
`hash` is a python builtin, so rather use `hash_` for example

magopian commented on validator/testcases/libraries.py:
```diff
@@ -0,0 +1,204 @@
+        if res and library in res:
+            return max_version(res[library])
+    else:
+        match = next(((k, v) for k, v in res.iteritems() if v),
```
This deserves a comment

magopian commented on validator/testcases/libraries.py:
```diff
@@ -0,0 +1,204 @@
+        match = next(((k, v) for k, v in res.iteritems() if v),
+                     None)
+        if match:
+            return match[0], max_version(match[1])
```
I would have written:
```python
matches = [(k, v) for k, v in res.iteritems() if v]
if matches:
    match = matches[0]
    return match[0], max_version(match[1])
```

Simpler, idiomatic, explicit, more readable.

magopian commented:
Thank you for your work, it is impressive.

In the future, please consider:
1/ multiple atomic commits
2/ multiple PRs if it's fixing unrelated things (eg: rewriting or improving some code to clean it is ok, completely rewriting huge blobs isn't)
3/ committing "file moves" and "code modifications" separately, to be able to only review the latter
4/ add comments and docstring as much as possible, this really helps understanding the code, reviewing it, and maintaining it in the future
5/ document the rationale, the changes... in the PR description. The discussion we had in person was very helpful for that (understanding the flow of lib updates, storing the hashes in olympia's database...)
6/ code simplicity. This is very important for maintainability. Code must be as clear, readable and straightforward as possible. Avoid complex and convoluted solutions at all costs, even if they are more concise or "more clever". In fact, avoid it in particular if they are.

Regarding the PR itself, as discussed, it would make sense to distribute the lib fetcher separately in the future, maybe even release it in a generic enough way so it may be useful to people outside of Mozilla?
Also, I don't think the code to update olympia's codebase with the hashes from the latest versions of the libs is in this PR, so if possible, i'd like:
1/ a separate bug/PR for the new lib updater code (not necessarily as a separate repo yet)
2/ a separate bug/PR to update olympia's database (using the web API if I remember correctly what you wanted to do from our discussion at Whistler?)
3/ maybe do not implement 2/ for now, as that would:
    - avoid having to wait for the fix of the oauth code on olympia
    - still let us run a cron job to take a local file and populate it in the DB (that would even be easier to implement)

magopian commented on validator/validate.py:
```diff
@@ -31,7 +31,9 @@ def validate(path, format="json",
         other format will simply return the error bundle.
     `approved_applications`:
-        Path to the list of approved application versions
+        Path to the list of approved application versions.
```
Document that it can be a dict (same for the `library_metadata` below).

Should we remove the `app_versions.json` file from the repository now that it's not used anymore? That would be convenient, as we won't have to maintain it manually anymore

kmaglione replied:
It's still used for now, when the validator is run standalone. Ideally, I'd like to be able to fetch it on demand from AMO, in the future, but at the moment we still need it.

kmaglione commented:
> 2/ multiple PRs if it's fixing unrelated things (eg: rewriting or improving some code to clean it is ok, completely rewriting huge blobs isn't)

I think the only unrelated things that snuck in here were in `installrdf.py`. I'm going to pull those changes out and put them into a separate PR.

> 3/ committing "file moves" and "code modifications" separately, to be able to only review the latter

I think there was only one file that actually moved here (`test_libraryblacklist.py`). The code from `jetpack.py` that I didn't delete actually got merged into `frameworks.py` in a way that I don't think rename tracking would have helped. I'll try to see if I can fix the former with a rebase.

> 6/ code simplicity. This is very important for maintainability. Code must be as clear, readable and straightforward as possible. Avoid complex and convoluted solutions at all costs, even if they are more concise or "more clever". In fact, avoid it in particular if they are.

I think you're going to point these out when you see them. I try to keep most of my code as simple as possible, so we probably have different definitions.

> 1/ a separate bug/PR for the new lib updater code (not necessarily as a separate repo yet)

I think I'll move them to a separate repo.

> 2/ a separate bug/PR to update olympia's database (using the web API if I remember correctly what you wanted to do from our discussion at Whistler?)

I have a separate repo for this, based on `andymckay/amo-oauth`. I need to do some cleanup, then I'll put it up somewhere for review.

> 3/ maybe do not implement 2/ for now, as that would:
> - avoid having to wait for the fix of the oauth code on olympia
> - still let us run a cron job to take a local file and populate it in the DB (that would even be easier to implement)

I'm not sure a cronjob be sufficient here, since we need to add manual hashes fairly often. It might be sufficient for the short term to split out the oauth changes and add a management command so someone from ops can update the hashes in the short term, though. I'd really rather not wait too long.

kmaglione commented:
I think this is everything.

I also removed the metadata updater and some unrelated changes that originally got sucked in, moved the metadata validation code that was into Olympia into `libraries.py` and `test_libraries.py`, and factored out some common code from a few test files.

kmaglione commented:
It turns out I forgot a couple of things, actually. I have an update which changes `"banned": (bool|string)` to `"deprecated": int`, and updates the README. It's marooned on my office computer at the moment, since it seems to have gotten kicked off the network. Will update later today.

kmaglione commented:
Looks like I forgot to push this last week. This is now mergeable again, and I think it addresses everything.

magopian commented:
A few comments aren't answered yet (eg https://github.com/mozilla/amo-validator/pull/271#discussion_r35101643 and https://github.com/mozilla/amo-validator/pull/271#discussion_r33466960).

Regarding your [last comment](https://github.com/mozilla/amo-validator/pull/271#issuecomment-119038083):
1/ you said "I think I'll move them to a separate repo.": is that done? If so, shouldn't the documentation (in the README, about the new data structure for libs/frameworks hashes) be in this new repo instead?
2/ your answer about the cronjob not being sufficient: I have a bad feeling about this, it really looks over-engineered.

I understand your need to have a simpler (not "blocked" by the filesystem/repo) way to update the hashes. In that case, the "usual" solution would be to create an admin model for that. Or a management command, yes.

This PR is really large, I can't remember what's in there, and what isn't, can you please sum it up? Is it "only" the code to use the new "lib/framework hashes data structure"? Does it still have the code related to the lib updating? Or the "posting to AMO to update its data"?

kmaglione commented:
> 1/ you said "I think I'll move them to a separate repo.": is that done?

Yes, but I haven't made it public yet.

> If so, shouldn't the documentation (in the README, about the new data structure for libs/frameworks hashes) be in this new repo instead?

I think it makes sense to have it in this repo, since the validator deals with it extensively.

> 2/ your answer about the cronjob not being sufficient: I have a bad feeling about this, it really looks over-engineered.
>
> I understand your need to have a simpler (not "blocked" by the filesystem/repo) way to update the hashes. In that case, the "usual" solution would be to create an admin model for that. Or a management command, yes.

That would be the usual solution, yes. The problem with that solution, though, is that all of the admin pages we currently have are kludgy and poorly-maintained. This one would need to deal with managing much more complex data, would probably have to be updated more often, would have to be updated in sync with the validator, would most likely have to include an incredible amount of logic for mapping DB models to data structures the validator deals with, would require a local tool anyway for hashing libraries and frameworks that need to be manually approved, would then require either uploading those new hashes to a running server and re-downloading the libraries database to test, or otherwise re-implementing the same logic again in a local client...

And, as for a management command, that would mean needlessly relying on someone from ops every time we need to add a hash, or change a flag.

In either case, we'd still need to implement all of the same logic. And on the server side, I think it would require much more complexity, not less.

At any rate, I'm not ruling out moving more of the logic to Olympia, but I'd prefer to go live with the current solution first, and talk about what needs to change if it doesn't appear to be working well.

> This PR is really large, I can't remember what's in there, and what isn't, can you please sum it up?

The current incarnation has:

* Code for detecting the following frameworks, and what files belong to the framework:

  - BestToolbars
  - Conduit
  - Crossrider
  - Jetpack (CFX)
  - Jetpack (JPM)
  - OpenForge
  - Kango
  - Scriptify

* New code for detecting known libraries with the added ability to:

  - Mark libraries, frameworks, or versions as banned.
  - Add warnings for particular libraries ("Prototype.js may not be used in browser overlays.").
  - Warn when a file which should belong to a framework is unknown.

The current version of the patch I have for AMO leaves out the API code for the moment, and just has an admin page with a file upload field. I think that should be sufficient in the short term.

In the medium term, we're going to need the API fixes in my original PR anyway if we're actually going to deliver on our promise of a signing API.

magopian commented:
Ok, here's the deal. This PR has grown so huge that I can't (and don't want to) go over it again. It includes so many different changes, some unrelated, that I just can't grok it.

On the other hand, I don't want to reject it "just because I'm too lazy to go over it (again)", and I know now (after some discussion with you on irc) that at least some of this work is needed for the future of the validator (and dealing with the chrome-style add-ons API). Also, you're the only current active maintainer, and you have agreed on a plan with management, if I understand things correctly.

So unless you find someone who's ready to review it, I'll just say "merge it while I'm looking the other way".

To be honest, I'm pretty annoyed at this whole thing, and would like this to not happen again in the future. If there's anything I can do to help this not happening again, please let me know.

muffinresearch commented:
I'm thinking if it would be useful to focus on capturing these changes on the addons-validator project - we're going to need to cover the same processes, but as there's clearly improvements to the amo-validator here it would be nice to get them rather than us just port what's there. 

A starting point might be to add some issues breaking down what's in this patch on the addons-validator project so we can address them as we bring across the JS lib/framework detection?

EnTeQuAk commented:
I'm going to close this for cleaning up reasons, not sure how relevant this PR still is. Feel free to reopen anytime though :)
